futures-core/src/task_impl/mod.rs

```diff
@@ -15,106 +15,64 @@ pub use self::std::*;
 #[cfg(not(feature = "std"))]
 pub use self::core::*;
 
-/// The context for the currently running task.
-pub struct Context {
-    waker: Option<Waker>,
+pub struct ContextInner<'a> {
+    unpark: TaskUnpark,
+    map: &'a LocalMap,
 }
 
-impl Context {
+/// The context for the currently running task.
+///
+/// A task represents a single lightweight "thread" of execution driving a
+/// future to completion.
+///
+/// In general, futures are composed into large units of work, which are then
+/// spawned as tasks onto an *executor*. The executor is responsible for polling
+/// the future as notifications arrive, until the future terminates.
+///
+/// The poll function accepts a context, which is usually provided by the executor.
+pub struct Context<'a>(Option<ContextInner<'a>>);
```

Comment:
no-op contexts seems to be supported by providing an empty `LocalMap` plus a no-op `NotifyHandle` (or collecting, for testing).

The use for `Context(None)` seems to be to implement the panicking context. Could this be accomplished with an empty `LocalMap` and panicking `NotifyHandle` instead to avoid the `Option`?

Reply:
Probably, though I don't think it's important to get it perfect in the first PR, since it's an implementation detail that can be changed locally at any time.